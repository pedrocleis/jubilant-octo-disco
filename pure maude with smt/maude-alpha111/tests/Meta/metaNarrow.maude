set show timing off .
set show advisories off .

***
***	Check handling of unification incompleteness.
***

mod FOO is
  sort Foo .
  op f : Foo Foo -> Foo [assoc] .
  ops g h : Foo -> Foo .
  ops a b c d e : -> Foo .

vars X Y Z W : Foo .
  rl g(f(X, X)) => h(X) [label nar].
endm

*** 4 unifiers + incompleteness
unify f(X, X) =? f(Y, Z, Z) .

*** metaNarrow()

red in META-LEVEL :
  metaNarrow(
    ['FOO],
    'g['f['Y:Foo, 'Z:Foo, 'Z:Foo]],
    'W:Foo,
    '+,
    unbounded,
    0) .

red in META-LEVEL :
  metaNarrow(
    ['FOO],
    'g['f['Y:Foo, 'Z:Foo, 'Z:Foo]],
    'W:Foo,
    '+,
    unbounded,
    1) .

red in META-LEVEL :
  metaNarrow(
    ['FOO],
    'g['f['Y:Foo, 'Z:Foo, 'Z:Foo]],
    'W:Foo,
    '+,
    unbounded,
    2) .

red in META-LEVEL :
  metaNarrow(
    ['FOO],
    'g['f['Y:Foo, 'Z:Foo, 'Z:Foo]],
    'W:Foo,
    '+,
    unbounded,
    3) .

red in META-LEVEL :
  metaNarrow(
    ['FOO],
    'g['f['Y:Foo, 'Z:Foo, 'Z:Foo]],
    'W:Foo,
    '+,
    unbounded,
    4) .

*** metaNarrowingApply()

red in META-LEVEL :
  metaNarrowingApply(
    ['FOO],
    'g['f['Y:Foo, 'Z:Foo, 'Z:Foo]],
    'nar,
    empty,
    '#,
    0) .

red in META-LEVEL :
  metaNarrowingApply(
    ['FOO],
    'g['f['Y:Foo, 'Z:Foo, 'Z:Foo]],
    'nar,
    empty,
    '#,
    1) .

red in META-LEVEL :
  metaNarrowingApply(
    ['FOO],
    'g['f['Y:Foo, 'Z:Foo, 'Z:Foo]],
    'nar,
    empty,
    '#,
    2) .

red in META-LEVEL :
  metaNarrowingApply(
    ['FOO],
    'g['f['Y:Foo, 'Z:Foo, 'Z:Foo]],
    'nar,
    empty,
    '#,
    3) .

red in META-LEVEL :
  metaNarrowingApply(
    ['FOO],
    'g['f['Y:Foo, 'Z:Foo, 'Z:Foo]],
    'nar,
    empty,
    '#,
    4) .

*** using # variables

red in META-LEVEL :
  metaNarrowingApply(
    ['FOO],
    'g['f['#22:Foo, '#23:Foo, '#23:Foo]],
    'nar,
    empty,
    '#,
    0) .

red in META-LEVEL :
  metaNarrowingApply(
    ['FOO],
    'g['f['#22:Foo, '#23:Foo, '#23:Foo]],
    'nar,
    empty,
    '#,
    1) .


red in META-LEVEL :
  metaNarrowingApply(
    ['FOO],
    'g['f['#22:Foo, '#23:Foo, '#23:Foo]],
    'nar,
    empty,
    '#,
    2) .

red in META-LEVEL :
  metaNarrowingApply(
    ['FOO],
    'g['f['#22:Foo, '#23:Foo, '#23:Foo]],
    'nar,
    empty,
    '#,
    3) .

red in META-LEVEL :
  metaNarrowingApply(
    ['FOO],
    'g['f['#22:Foo, '#23:Foo, '#23:Foo]],
    'nar,
    empty,
    '#,
    4) .

***
***	Now with the incompleteness arising from a variant equation.
***

mod BAR is
  sort Foo .
  op f : Foo Foo -> Foo [assoc] .
  ops g h i : Foo -> Foo .
  ops a b c d e : -> Foo .

vars X Y Z W : Foo .
  eq g(f(X, X)) = h(X) [variant].
  rl h(X) => i(X) [label nar] .
endm

variant unify h(X) =? g(f(Y, Z, Z)) .

red in META-LEVEL :
  metaNarrowingApply(
    ['BAR],
    'g['f['Y:Foo, 'Z:Foo, 'Z:Foo]],
    'nar,
    empty,
    '#,
    0) .

red in META-LEVEL :
  metaNarrowingApply(
    ['BAR],
    'g['f['Y:Foo, 'Z:Foo, 'Z:Foo]],
    'nar,
    empty,
    '#,
    1) .

red in META-LEVEL :
  metaNarrowingApply(
    ['BAR],
    'g['f['Y:Foo, 'Z:Foo, 'Z:Foo]],
    'nar,
    empty,
    '#,
    2) .

red in META-LEVEL :
  metaNarrowingApply(
    ['BAR],
    'g['f['Y:Foo, 'Z:Foo, 'Z:Foo]],
    'nar,
    empty,
    '#,
    3) .

red in META-LEVEL :
  metaNarrowingApply(
    ['BAR],
    'g['f['Y:Foo, 'Z:Foo, 'Z:Foo]],
    'nar,
    empty,
    '#,
    4) .
