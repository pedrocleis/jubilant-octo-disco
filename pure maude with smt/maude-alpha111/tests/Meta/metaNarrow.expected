Maude> ==========================================
unify in FOO : f(X, X) =? f(Y, Z, Z) .

Solution 1
X --> f(#1:Foo, #2:Foo, #2:Foo, #1:Foo, #2:Foo)
Y --> f(#1:Foo, #2:Foo, #2:Foo, #1:Foo)
Z --> f(#2:Foo, #1:Foo, #2:Foo)
Warning: Unification modulo the theory of operator f has encountered an
    instance for which it may not be complete.

Solution 2
X --> f(#1:Foo, #1:Foo, #1:Foo)
Y --> f(#1:Foo, #1:Foo)
Z --> f(#1:Foo, #1:Foo)

Solution 3
X --> f(#1:Foo, #2:Foo, #2:Foo)
Y --> f(#1:Foo, #2:Foo, #2:Foo, #1:Foo)
Z --> #2:Foo

Solution 4
X --> f(#1:Foo, #1:Foo)
Y --> f(#1:Foo, #1:Foo)
Z --> #1:Foo
Warning: Some unifiers may have been missed due to incomplete unification
    algorithm(s).
==========================================
reduce in META-LEVEL : metaNarrow(['FOO], 'g['f['Y:Foo,'Z:Foo,'Z:Foo]], 'W:Foo,
    '+, unbounded, 0) .
rewrites: 2
result ResultTriple: {'h['f['%1:Foo,'%2:Foo,'%2:Foo,'%1:Foo,'%2:Foo]],'Foo,
	'W:Foo <- 'h['f['%1:Foo,'%2:Foo,'%2:Foo,'%1:Foo,'%2:Foo]]}
==========================================
reduce in META-LEVEL : metaNarrow(['FOO], 'g['f['Y:Foo,'Z:Foo,'Z:Foo]], 'W:Foo,
    '+, unbounded, 1) .
Warning: Unification modulo the theory of operator f has encountered an
    instance for which it may not be complete.
rewrites: 2
result ResultTriple: {'h['f['%1:Foo,'%1:Foo,'%1:Foo]],'Foo,
	'W:Foo <- 'h['f['%1:Foo,'%1:Foo,'%1:Foo]]}
==========================================
reduce in META-LEVEL : metaNarrow(['FOO], 'g['f['Y:Foo,'Z:Foo,'Z:Foo]], 'W:Foo,
    '+, unbounded, 2) .
rewrites: 2
result ResultTriple: {'h['f['%1:Foo,'%2:Foo,'%2:Foo]],'Foo,
	'W:Foo <- 'h['f['%1:Foo,'%2:Foo,'%2:Foo]]}
==========================================
reduce in META-LEVEL : metaNarrow(['FOO], 'g['f['Y:Foo,'Z:Foo,'Z:Foo]], 'W:Foo,
    '+, unbounded, 3) .
rewrites: 2
result ResultTriple: {'h['f['%1:Foo,'%1:Foo]],'Foo,
	'W:Foo <- 'h['f['%1:Foo,'%1:Foo]]}
==========================================
reduce in META-LEVEL : metaNarrow(['FOO], 'g['f['Y:Foo,'Z:Foo,'Z:Foo]], 'W:Foo,
    '+, unbounded, 4) .
rewrites: 2
result ResultTriple?: (failureIncomplete).ResultTriple?
==========================================
reduce in META-LEVEL : metaNarrowingApply(['FOO], 'g['f['Y:Foo,'Z:Foo,'Z:Foo]],
    'nar, empty, '#, 0) .
Warning: Unification modulo the theory of operator f has encountered an
    instance for which it may not be complete.
rewrites: 3
result NarrowingResult: {'h['f['@1:Foo,'@2:Foo,'@2:Foo,'@1:Foo,'@2:Foo]],'Foo,
	'Y:Foo <- 'f['@1:Foo,'@2:Foo,'@2:Foo,'@1:Foo] ; 
	'Z:Foo <- 'f['@2:Foo,'@1:Foo,'@2:Foo],
	'X:Foo <- 'f['@1:Foo,'@2:Foo,'@2:Foo,'@1:Foo,'@2:Foo],
[],'@}
==========================================
reduce in META-LEVEL : metaNarrowingApply(['FOO], 'g['f['Y:Foo,'Z:Foo,'Z:Foo]],
    'nar, empty, '#, 1) .
rewrites: 3
result NarrowingResult: {'h['f['@1:Foo,'@1:Foo,'@1:Foo]],'Foo,
	'Y:Foo <- 'f['@1:Foo,'@1:Foo] ; 
	'Z:Foo <- 'f['@1:Foo,'@1:Foo],
	'X:Foo <- 'f['@1:Foo,'@1:Foo,'@1:Foo],
[],'@}
==========================================
reduce in META-LEVEL : metaNarrowingApply(['FOO], 'g['f['Y:Foo,'Z:Foo,'Z:Foo]],
    'nar, empty, '#, 2) .
rewrites: 3
result NarrowingResult: {'h['f['@1:Foo,'@2:Foo,'@2:Foo]],'Foo,
	'Y:Foo <- 'f['@1:Foo,'@2:Foo,'@2:Foo,'@1:Foo] ; 
	'Z:Foo <- '@2:Foo,
	'X:Foo <- 'f['@1:Foo,'@2:Foo,'@2:Foo],
[],'@}
==========================================
reduce in META-LEVEL : metaNarrowingApply(['FOO], 'g['f['Y:Foo,'Z:Foo,'Z:Foo]],
    'nar, empty, '#, 3) .
rewrites: 3
result NarrowingResult: {'h['f['@1:Foo,'@1:Foo]],'Foo,
	'Y:Foo <- 'f['@1:Foo,'@1:Foo] ; 
	'Z:Foo <- '@1:Foo,
	'X:Foo <- 'f['@1:Foo,'@1:Foo],
[],'@}
==========================================
reduce in META-LEVEL : metaNarrowingApply(['FOO], 'g['f['Y:Foo,'Z:Foo,'Z:Foo]],
    'nar, empty, '#, 4) .
rewrites: 2
result NarrowingResult?: (failureIncomplete).NarrowingResult?
==========================================
reduce in META-LEVEL : metaNarrowingApply(['FOO], 'g['f['#22:Foo,'#23:Foo,
    '#23:Foo]], 'nar, empty, '#, 0) .
Warning: Unification modulo the theory of operator f has encountered an
    instance for which it may not be complete.
rewrites: 3
result NarrowingResult: {'h['f['@1:Foo,'@2:Foo,'@2:Foo,'@1:Foo,'@2:Foo]],'Foo,
	'#22:Foo <- 'f['@1:Foo,'@2:Foo,'@2:Foo,'@1:Foo] ; 
	'#23:Foo <- 'f['@2:Foo,'@1:Foo,'@2:Foo],
	'X:Foo <- 'f['@1:Foo,'@2:Foo,'@2:Foo,'@1:Foo,'@2:Foo],
[],'@}
==========================================
reduce in META-LEVEL : metaNarrowingApply(['FOO], 'g['f['#22:Foo,'#23:Foo,
    '#23:Foo]], 'nar, empty, '#, 1) .
rewrites: 3
result NarrowingResult: {'h['f['@1:Foo,'@1:Foo,'@1:Foo]],'Foo,
	'#22:Foo <- 'f['@1:Foo,'@1:Foo] ; 
	'#23:Foo <- 'f['@1:Foo,'@1:Foo],
	'X:Foo <- 'f['@1:Foo,'@1:Foo,'@1:Foo],
[],'@}
==========================================
reduce in META-LEVEL : metaNarrowingApply(['FOO], 'g['f['#22:Foo,'#23:Foo,
    '#23:Foo]], 'nar, empty, '#, 2) .
rewrites: 3
result NarrowingResult: {'h['f['@1:Foo,'@2:Foo,'@2:Foo]],'Foo,
	'#22:Foo <- 'f['@1:Foo,'@2:Foo,'@2:Foo,'@1:Foo] ; 
	'#23:Foo <- '@2:Foo,
	'X:Foo <- 'f['@1:Foo,'@2:Foo,'@2:Foo],
[],'@}
==========================================
reduce in META-LEVEL : metaNarrowingApply(['FOO], 'g['f['#22:Foo,'#23:Foo,
    '#23:Foo]], 'nar, empty, '#, 3) .
rewrites: 3
result NarrowingResult: {'h['f['@1:Foo,'@1:Foo]],'Foo,
	'#22:Foo <- 'f['@1:Foo,'@1:Foo] ; 
	'#23:Foo <- '@1:Foo,
	'X:Foo <- 'f['@1:Foo,'@1:Foo],
[],'@}
==========================================
reduce in META-LEVEL : metaNarrowingApply(['FOO], 'g['f['#22:Foo,'#23:Foo,
    '#23:Foo]], 'nar, empty, '#, 4) .
rewrites: 2
result NarrowingResult?: (failureIncomplete).NarrowingResult?
==========================================
variant unify in BAR : h(X) =? g(f(Y, Z, Z))  .

Warning: Unification modulo the theory of operator f has encountered an
    instance for which it may not be complete.
Unifier #1
rewrites: 4
X --> f(#1:Foo, #2:Foo, #2:Foo, #1:Foo, #2:Foo)
Y --> f(#1:Foo, #2:Foo, #2:Foo, #1:Foo)
Z --> f(#2:Foo, #1:Foo, #2:Foo)

Unifier #2
rewrites: 4
X --> f(#1:Foo, #1:Foo, #1:Foo)
Y --> f(#1:Foo, #1:Foo)
Z --> f(#1:Foo, #1:Foo)

Unifier #3
rewrites: 4
X --> f(#1:Foo, #2:Foo, #2:Foo)
Y --> f(#1:Foo, #2:Foo, #2:Foo, #1:Foo)
Z --> #2:Foo

Unifier #4
rewrites: 4
X --> f(#1:Foo, #1:Foo)
Y --> f(#1:Foo, #1:Foo)
Z --> #1:Foo

No more unifiers.
Warning: Some unifiers may have been missed due to incomplete unification
    algorithm(s).
rewrites: 4
==========================================
reduce in META-LEVEL : metaNarrowingApply(['BAR], 'g['f['Y:Foo,'Z:Foo,'Z:Foo]],
    'nar, empty, '#, 0) .
Warning: Unification modulo the theory of operator f has encountered an
    instance for which it may not be complete.
rewrites: 7
result NarrowingResult: {'i['f['%1:Foo,'%2:Foo,'%2:Foo,'%1:Foo,'%2:Foo]],'Foo,
	'Y:Foo <- 'f['%1:Foo,'%2:Foo,'%2:Foo,'%1:Foo] ; 
	'Z:Foo <- 'f['%2:Foo,'%1:Foo,'%2:Foo],
	'X:Foo <- 'f['%1:Foo,'%2:Foo,'%2:Foo,'%1:Foo,'%2:Foo],
[],'%}
==========================================
reduce in META-LEVEL : metaNarrowingApply(['BAR], 'g['f['Y:Foo,'Z:Foo,'Z:Foo]],
    'nar, empty, '#, 1) .
rewrites: 3
result NarrowingResult: {'i['f['%1:Foo,'%1:Foo,'%1:Foo]],'Foo,
	'Y:Foo <- 'f['%1:Foo,'%1:Foo] ; 
	'Z:Foo <- 'f['%1:Foo,'%1:Foo],
	'X:Foo <- 'f['%1:Foo,'%1:Foo,'%1:Foo],
[],'%}
==========================================
reduce in META-LEVEL : metaNarrowingApply(['BAR], 'g['f['Y:Foo,'Z:Foo,'Z:Foo]],
    'nar, empty, '#, 2) .
rewrites: 3
result NarrowingResult: {'i['f['%1:Foo,'%2:Foo,'%2:Foo]],'Foo,
	'Y:Foo <- 'f['%1:Foo,'%2:Foo,'%2:Foo,'%1:Foo] ; 
	'Z:Foo <- '%2:Foo,
	'X:Foo <- 'f['%1:Foo,'%2:Foo,'%2:Foo],
[],'%}
==========================================
reduce in META-LEVEL : metaNarrowingApply(['BAR], 'g['f['Y:Foo,'Z:Foo,'Z:Foo]],
    'nar, empty, '#, 3) .
rewrites: 3
result NarrowingResult: {'i['f['%1:Foo,'%1:Foo]],'Foo,
	'Y:Foo <- 'f['%1:Foo,'%1:Foo] ; 
	'Z:Foo <- '%1:Foo,
	'X:Foo <- 'f['%1:Foo,'%1:Foo],
[],'%}
==========================================
reduce in META-LEVEL : metaNarrowingApply(['BAR], 'g['f['Y:Foo,'Z:Foo,'Z:Foo]],
    'nar, empty, '#, 4) .
rewrites: 2
result NarrowingResult?: (failureIncomplete).NarrowingResult?
Maude> Bye.
